proxy/src/telemetry/metrics/latency.rs

```diff
@@ -0,0 +1,285 @@
+use std::cell::Cell;
+use std::default::Default;
+use std::time::Duration;
+
+/// A series of latency values and counts.
+#[derive(Debug)]
+pub struct Latencies(Buckets<Latency, Buckets<Latency>>);
+
+/// A latency in tenths of a millisecond.
+#[derive(Debug, Default, Eq, PartialEq, Ord, PartialOrd, Copy, Clone, Hash)]
+pub struct Latency(u32);
+
+/// Generalization of a histogram bucket counting observations of type `T`.
+trait Bucket<T: PartialOrd> {
+    /// Returns `true` if a given `observation` falls into this `Bucket`.
+    fn contains(&self, observation: &T) -> bool;
+    /// Observe a value, incrementing this bucket if it `contains` that value.
+    /// 
+    /// # Returns
+    /// - `true` if this `Bucket`'s count was incremented;
+    /// - `false` otherwise.
+    fn add(&self, observation: &T) -> bool;
+}
+
+/// A "leaf" bucket containing a single counter.
+/// 
+/// # Type Parameters:
+/// - `T`: the type of observations counted by this bucket.
+/// - `C`: the type used for counting observed values. Bounded by `Copy`
+///        as it is placed in a `Cell`; defaults to `u32`.
+#[derive(Debug, Clone)]
+struct LeafBucket<T, C: Copy = u32> {
+    /// The upper bound of the range of values in this bucket.
+    max: T,
+    /// Count of observations falling into this bucket.
+    /// 
+    /// `Cell` is used for interior mutability here, as it should be possible
+    /// to increment the count in a bucket without mutably locking the entire 
+    /// histogram structure containing this `LeafBucket`; as would be necessary to
+    /// change the _shape_ of the histogram.
+    // XXX should this be an atomic integer since it goes in a `Cell`?
+    count: Cell<C>,
+}
+
+
+/// An array of [`Bucket`]s which can themselves be a bucket.
+/// 
+/// [`Bucket`]: trait.Bucket.html
+#[derive(Debug, Clone)]
+struct Buckets<T: PartialOrd, B: Bucket<T>=LeafBucket<T>> {
+    /// The upper bound of this `Bucket`.
+    /// 
+    /// This should be the highest `max` value of the buckets herein contained.
+    max: T,
+    /// The aforementioned `Bucket`s herein contained.
+    /// 
+    /// Currently, these must be in order.
+    // FIXME: number of buckets per bucket set is hardcoded currently.
+    buckets: [B; 5]
+}
+
+impl<T> LeafBucket<T> {
+    // Construct a new `LeafBucket` with the upper bound of `max`.
+    #[inline] fn new<I: Into<T>>(max: I) -> Self {
+        LeafBucket {
+            max: max.into(),
+            count: Cell::new(0)
+        }
+
+    }
+}
+
+impl<T> Bucket<T> for LeafBucket<T>
+where T: PartialOrd {
+    fn contains(&self, t: &T) -> bool {
+        t <= &self.max
+    }
+
+    fn add(&self, t: &T) -> bool {
+        if self.contains(t) {
+            self.count.set(self.count.get().saturating_add(1));
+            true
+        } else {
+            false
+        }
+
+    }
+}
+
+
+impl<T, B> Bucket<T> for Buckets<T, B>
+where
+    T: PartialOrd,
+    B: Bucket<T>,
+{
+    #[inline] fn contains(&self, t: &T) -> bool {
```

Comment:
nit: prefer a newline after attributes.

Though, most of the `inline`s aren't really needed, since the hint is mostly just to allow inline across crates, and all these functions are only used in the same crate, so LLVM can already decide if they should be inlined.